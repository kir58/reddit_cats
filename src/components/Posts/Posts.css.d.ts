// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'author': string;
  'date': string;
  'favourit': string;
  'favourites_icon': string;
  'flair_text': string;
  'item': string;
  'item_header': string;
  'item_header_element': string;
  'list': string;
  'loader_by_center': string;
  'picture': string;
  'video': string;
  'video_container': string;
}
export const cssExports: CssExports;
export default cssExports;
