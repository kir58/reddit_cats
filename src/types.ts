export type CatDetail = {
  data: {
    id: string,
    name: string,
    created_utc: number,
    title: string,
    author: string,
    link_flair_text: string,
    post_hint: string,
    url: string, 
    media: { reddit_video: {  fallback_url: string } },
  },
  kind: string,
}

export type CatsIds = Array<string>;