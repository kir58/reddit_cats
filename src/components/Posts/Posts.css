.list {
  list-style: none;
  padding: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  flex-wrap: wrap;
  width: 100%;
  
}
.item {
  border: 1px solid #002233;
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  margin: 20px 0;
  padding: 30px;
  width: 80%;
  background-color: #fafafa;
}
.item:hover {
  box-shadow: 0 10px 12px -8px green;
}

.item_header{
  display: flex;
  flex-wrap: wrap;
  color: #2f353b;
  font-size: 16px;
  padding: 10px 0;
  line-height: 22px;
}

.item_header_element {
  margin-right: 5px;
  margin-top: 5px;
}

.date {
  background: #e9c5bc;
  border-radius: 5px;
  padding: 0 5px;
}

.favourit {
	display: none;
}

.favourites_icon {
  height: 20px;
  cursor: pointer;
  margin-right: 5px;
  margin-top: 5px;
}
.video_container {
  background-color: #000000;
  align-self: center;
  width: 100%;
  display: flex;
  justify-content: center;
  margin: 25px 0;
}
.video {
  max-width: 100%; 
  max-height: 512px;
}
.picture {
  max-height: 512px;
  max-width: 100%;
  margin: 25px 0;
  align-self: center;
}

.flair_text {
  background-color: #e6e1e1;
  font-size: 18px;
  padding: 3px 5px;
  border-radius: 7px;
  display: inline-block;
}

.author {
  text-decoration: underline;
}

.loader_by_center {
  position: absolute;
  top: 50%;
  left: 50%;
}

@media screen and (max-width: 480px) {
  .video {
    max-height: 360px;
  }

  .picture {
    max-height: 360px;
  }
}