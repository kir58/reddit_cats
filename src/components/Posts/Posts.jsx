/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';

import PropTypes from 'prop-types';
import styles from './Posts.css';
import { getDiffDate } from '../../utils';
import FullStar from '../../assest/full_star.png';
import StarWithoutBackground from '../../assest/star_without_background.png';


const Posts = (props) => {
  const { cats, favouritesCatsIds, handleToggle } = props;
  return (
    <ul className={styles.list}>
      {cats.map(({ data }) => (
        <li key={data.id} className={styles.item}>
          <div className={styles.item_header}>
            <button
              id={data.name}
              name={data.name}
              aria-label={data.name}
              type="button"
              className={styles.favourit}
              onClick={handleToggle(data.name)}
            />
            <label htmlFor={data.name}>
              <img
                src={favouritesCatsIds.includes(data.name) ? FullStar : StarWithoutBackground}
                alt="favourites"
                className={styles.favourites_icon}
                title={favouritesCatsIds.includes(data.name) ? 'remove from favourites' : 'add to favourites'}
              />
            </label>
            <div className={styles.item_header_element}>
              {'Posted by '}
              <span className={`${styles.author} ${styles.item_header_element}`}>{data.author}</span>
            </div>
            <div className={`${styles.date} ${styles.item_header_element}`}>{` ${getDiffDate(data.created_utc)}`}</div>
          </div>
          <div>
            {`${data.title} `}
            <span className={styles.flair_text}>{data.link_flair_text}</span>
          </div>
          {data.post_hint === 'image' ? <img src={data.url} alt="cat" className={styles.picture} /> : null}
          {data.post_hint === 'hosted:video' ? (
            <div className={styles.video_container}>
              <video src={data.media.reddit_video.fallback_url} className={styles.video} controls="controls">
                <track src="#" kind="captions" srcLang="en" label="english_captions" />
              </video>
            </div>
          ) : null}
        </li>
      ))}
    </ul>
  );
};

Posts.propTypes = {
  cats: PropTypes.array.isRequired,
  favouritesCatsIds: PropTypes.array.isRequired,
  handleToggle: PropTypes.func.isRequired,
};

export default Posts;
