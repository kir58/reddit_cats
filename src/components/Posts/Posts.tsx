import React from 'react';

import styles from './Posts.css';
import { getDiffDate } from '../../utils';
import FullStar from '../../assest/full_star.png';
import StarWithoutBackground from '../../assest/star_without_background.png';
import { CatDetail, CatsIds } from '../../types';



type Props = {
  onToggle: Function,
  favouritesCatsIds: CatsIds,
  cats: CatDetail[],

}
const Posts = ({ cats, favouritesCatsIds, onToggle }: Props) =>  (
  <ul className={styles.list}>
    {cats.map(({ data: { id, name, author, created_utc, title, link_flair_text, post_hint, url, media, } }) => (
      <li key={id} className={styles.item}>
        <div className={styles.item_header}>
          <button
            id={name}
            name={name}
            aria-label={name}
            type="button"
            className={styles.favourit}
            onClick={onToggle(name)}
          />
          <label htmlFor={name}>
            <img
              src={favouritesCatsIds.includes(name) ? FullStar : StarWithoutBackground}
              alt="favourites"
              className={styles.favourites_icon}
              title={favouritesCatsIds.includes(name) ? 'remove from favourites' : 'add to favourites'}
            />
          </label>
          <div className={styles.item_header_element}>
            {'Posted by '}
            <span className={`${styles.author} ${styles.item_header_element}`}>{author}</span>
          </div>
          <div className={`${styles.date} ${styles.item_header_element}`}>{` ${getDiffDate(created_utc)}`}</div>
        </div>
        <div>
          {`${title} `}
          <span className={styles.flair_text}>{link_flair_text}</span>
        </div>
        {post_hint === 'image' ? <img src={url} alt="cat" className={styles.picture} /> : null}
        {post_hint === 'hosted:video' ? (
          <div className={styles.video_container}>
            <video src={media.reddit_video.fallback_url} className={styles.video} controls>
              <track src="#" kind="captions" srcLang="en" label="english_captions" />
            </video>
          </div>
        ) : null}
      </li>
    ))}
  </ul>
);

export default Posts;
