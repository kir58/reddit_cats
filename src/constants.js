export const LIMIT = 6;
export const POST_HEIGHT = 512;
