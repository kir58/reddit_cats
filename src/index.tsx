/* eslint-disable no-undef */
import '@babel/polyfill';
import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';
import App from './components/App/App';
import reducers from './reducers';

const favouritesCatsIdsFromLocalStorage = localStorage.getItem('favouritesCatsIds');
const favouritesCatsIds = {
  favouritesCatsIds: favouritesCatsIdsFromLocalStorage ? JSON.parse(favouritesCatsIdsFromLocalStorage) : [],
};

const store = createStore(
  reducers,
  favouritesCatsIds,
  composeWithDevTools(
    applyMiddleware(thunk),
  ),
);

store.subscribe(() => {
  const { favouritesCatsIds } = store.getState();
  localStorage.setItem('favouritesCatsIds', JSON.stringify(favouritesCatsIds));
});

render(<Provider store={store}><App /></Provider>, document.getElementById('root'));
