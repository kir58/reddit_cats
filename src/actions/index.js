import { createAction } from 'redux-actions';
import axios from 'axios';
import { LIMIT } from '../constants';

export const fetchPostsRequest = createAction('POSTS_FETCH_REQUEST');
export const fetchPostsSuccess = createAction('POSTS_FETCH_SUCCESS');
export const fetchPostsFailure = createAction('POSTS_FETCH_FAILURE');
export const toggleFavouritePosts = createAction('TOGGLE_FAVOURITE_POSTS');

export const fetchPosts = (lastId = '') => async (dispatch) => {
  dispatch(fetchPostsRequest());
  try {
    const response = await axios.get(`https://www.reddit.com/r/cats.json?after=${lastId}&limit=${LIMIT}`);
    dispatch(fetchPostsSuccess({ cats: response.data.data.children }));
  } catch (e) {
    dispatch(fetchPostsFailure());
  }
};
