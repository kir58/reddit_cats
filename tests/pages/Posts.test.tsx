import React from "react";
import { shallowToJson } from 'enzyme-to-json';
import Posts from '../../src/components/Posts/Posts';
import { PostsProps } from "../../src/types";
import { shallow } from '../setup/test-setup';
import { catsData } from "../fixtures/cats";

describe('<Posts />', () => {
  let wrapper : any;

  const props: PostsProps = {
    cats: catsData,
    favouritesCatsIds: ['cat3'],
    onToggle: jest.fn(),
  };

  beforeEach(() => {
    wrapper = shallow(<Posts {...props} />);
  });

  it('should match the snapshot', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
})