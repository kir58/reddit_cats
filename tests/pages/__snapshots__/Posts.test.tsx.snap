// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Posts /> should match the snapshot 1`] = `
<ul>
  <li
    key="1"
  >
    <div>
      <button
        aria-label="cat1"
        id="cat1"
        name="cat1"
        type="button"
      />
      <label
        htmlFor="cat1"
      >
        <img
          alt="favourites"
          src="test-file-stub"
          title="add to favourites"
        />
      </label>
      <div>
        Posted by 
        <span
          className="undefined undefined"
        >
          master cat
        </span>
      </div>
      <div
        className="undefined undefined"
      >
         just now
      </div>
    </div>
    <div>
      cat singer 
      <span>
        Cat Picture
      </span>
    </div>
    <img
      alt="cat"
      src="https://i.redd.it/xy7k06kx4py51.jpg"
    />
  </li>
  <li
    key="2"
  >
    <div>
      <button
        aria-label="cat2"
        id="cat2"
        name="cat2"
        type="button"
      />
      <label
        htmlFor="cat2"
      >
        <img
          alt="favourites"
          src="test-file-stub"
          title="add to favourites"
        />
      </label>
      <div>
        Posted by 
        <span
          className="undefined undefined"
        >
          master cat
        </span>
      </div>
      <div
        className="undefined undefined"
      >
         just now
      </div>
    </div>
    <div>
      cat singer 
      <span>
        Cat Picture
      </span>
    </div>
    <img
      alt="cat"
      src="https://i.redd.it/xy7k06kx4py51.jpg"
    />
  </li>
  <li
    key="3"
  >
    <div>
      <button
        aria-label="cat3"
        id="cat3"
        name="cat3"
        type="button"
      />
      <label
        htmlFor="cat3"
      >
        <img
          alt="favourites"
          src="test-file-stub"
          title="remove from favourites"
        />
      </label>
      <div>
        Posted by 
        <span
          className="undefined undefined"
        >
          master cat
        </span>
      </div>
      <div
        className="undefined undefined"
      >
         just now
      </div>
    </div>
    <div>
      cat singer 
      <span>
        Cat Picture
      </span>
    </div>
    <div>
      <video
        controls={true}
        src="https://v.redd.it/gorcce379sy51/DASH_1080.mp4?source=fallback"
      >
        <track
          kind="captions"
          label="english_captions"
          src="#"
          srcLang="en"
        />
      </video>
    </div>
  </li>
</ul>
`;
